### PR TITLE
HHH-11543 - ForeignKey to TABLE_PER_CLASS should target table if only one exists

### PR DESCRIPTION
**Pull request by beikov**
https://hibernate.atlassian.net/browse/HHH-11543

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/boot/internal/InFlightMetadataCollectorImpl.java [hunk 1]
```diff
@@ -1809,7 +1809,17 @@ protected void secondPassCompileForeignKeys(
 					secondPassCompileForeignKeys( referencedClass.getSuperclass().getTable(), done, buildingContext );
 				}
 
-				fk.setReferencedTable( referencedClass.getTable() );
+				if ( referencedClass.isAbstract() && referencedClass.getSubclassSpan() == 1 ) {
+					Iterator<PersistentClass> iter = referencedClass.getSubclassClosureIterator();
```
Comment (dreab8) [inline hibernate-core/src/main/java/org/hibernate/boot/internal/InFlightMetadataCollectorImpl.java]: referencedClass.isAbstract() may be null so if( referencedClass.isAbstract() ... can cause a NPE
I would change the code to 
if ( referencedClass.isAbstract() != null && referencedClass.isAbstract() && referencedClass.getSubclassSpan() == 1 ) {

**Comment by dreab8** [comment]
@beikov 
The fix causes also the failure of org.hibernate.test.hql.WithClauseJoinRewriteTest

**Comment by beikov** [comment]
@dreab8 I updated it, tests should pass now

**Comment by dreab8** [comment]
@beikov 
I think that creating a FK may cause problems, if at some point a new 
```
@Entity
	@Table(name = "SUPER_USER")
	public class SuperUser extends User {

	}
```
is introduced without dropping and recreating the entire db, the generated FK will prevent to add a SuperUser as a UserTask#user. 
Am I right?

**Comment by beikov** [comment]
You wouldn't have to drop the entire DB but only the FK, but essentially you are right, if a new type is added, the FK will become _invalid_.

**Comment by dreab8** [comment]
I think the FK should be generated only for the concrete class in this case `RegularUser regularUser`
as it happens with the current code.

**Comment by beikov** [comment]
Ok so reject the issue then?

**Comment by dreab8** [comment]
I'm for rejecting it.
@sebersole @Naros @gbadner what do u think?

**Comment by dreab8** [comment]
anyway looking at the issue, it seems the problem may be in the schema update not creating the FK to the RegularUser, it seems the schema creation creates it.

**Comment by sebersole** [comment]
IMO those should be consistent outcomes.  Either schema-create ignores it or schema-update generates it.

**Comment by dreab8** [comment]
running the same code of the issue and also the schema update generates the
```
 alter table UserTask 
       add constraint FKn8m7j58j0tytuc6breqixuga 
       foreign key (UserID_FK) 
       references RegularUsers
```

**Comment by dreab8** [comment]
I managed to reproduce the issue, 
if we have 

    ```
 @Table(name = "Users")
	@Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
	public static abstract class User{
            ...
      }

      @Entity
	@Table(name = "RegularUsers")
	public class RegularUser extends User {
	}

      @Entity
	@Table(name = "Companies")
	public class Company extends User {
	}

      @Entity
	@Table(name = "UserTask")
	public static class UserTask implements Serializable {

		...

		@ManyToOne(cascade = CascadeType.ALL, fetch = FetchType.LAZY)
		@JoinColumn(name = "UserID_FK", referencedColumnName = "UserID")
		private RegularUser regularUser;

		...
	}
```
      a FK is generated
```
alter table UserTask 
       add constraint FKn8m7j58j0tytuc6breqixuga 
       foreign key (UserID_FK) 
       references RegularUsers
```

when we add

```
 @Entity
	@Table(name = "Admins")
	public class Admin extends RegularUser {
		String adminField;
	}
```
the FK to the RegulaUser table is not generate anymore but I think this is correct, now `UserTask#regularUser` can contain instances of `Admin` that has his own table.

**Comment by beikov** [comment]
That is indeed correct and is also what I tried to explain to the user in the comments of the issue.

**Pull request closed without merging**